### DISCUSSION TITLE
include array index in `EvFilter` output.

### DISCUSSION THREAD

**Discussion by ryanleecode** [polkadot-api/polkadot-api#642] [General]
When doing a `force_batch` call, I want to check which items failed. 

Currently I am doing this.

```ts
const result = await firstValueFrom(
      peopleApi.tx.Utility.force_batch({ ... }))
const failures = peopleApi.event.Utility.ItemFailed.filter(result.events)
```

But i have a problem because the [`ItemFailed` event](https://github.com/paritytech/polkadot-sdk/blob/a4b51593dbe9b9390fa64f5040a9230b5d33b898/substrate/frame/utility/src/lib.rs#L120) doesn't expose index of the failure like `BatchInterrupted` does.

So to cross reference the failures with the actual index, I am doing this.
```ts
const failures = result.events
  .filter(
    (e) =>
      e.type === 'Utility' &&
      (e.value.type === 'ItemFailed' || e.value.type === 'ItemCompleted'),
  )
  .map((e, i) => ({ index: i, event: e }))
  .filter((e) => e.event.value.type === 'ItemFailed')
  .map((e) => ({ event: peopleApi.event.Utility.ItemFailed.filter([e.event])[0], index: e.index }))
   
```

This seems like a common usecase; hence, this is a proposal to include the array index in the output of `EvFilter` or create a separate function that does.

**Comment by voliva** [comment]
I like the idea, but I can't think of any non-breaking way where the API would work for everyone. And given that it can be done separately we may have to leave it as it is.

The problem is that the value `EventEntry.filter` returns is an array of the values of that event. I think this is good because if you've filtered for one specific event, you don't care anymore about the `{ type }` of that event, since it's already been set, and you get the value directly.

But now this value can be anything. If it was guaranteed to be an object with a specific shape, then we could add an `index` property. But we cannot because if it's a primitive or the value already has an `index` prop, then we would be overriding that.

I also don't fully understand your example - You are pre-filtering the events, then setting the index, then filtering again. The index you receive will be the one after the first filter, so I'm not sure how that helps.

Knowing that the result comes back in the same order as the original events, it should be fairly simple to write a utility to get the indices in linear time. Something like this should work:

```ts
type SystemEvent = { type: string; value: { type: string; value: unknown } }
function getIndicies<T>(events: Array<SystemEvent>, values: Array<T>) {
  let nextIdx = 0;
  const findNext = (value: T) => {
    while (nextIdx < events.length && events[nextIdx] !== value) nextIdx++;
    return nextIdx++;
  };

  return values.map((value) => ({ index: findNext(value), value }));
}

const filteredEvents = peopleApi.event.Utility.ItemFailed.filter(result.events);
const eventsWithIndex = getIndicies(result.events, filteredEvents)
```

**Comment by ryanleecode** [comment]
ideally there would be a new function that returns both the value and the index as a tuple of size 2. 

~~Alternatively you can just expose the type guard used in the filter and then i can easily just do `myArray.filter((x, i) => [papiSuppliedGuard(x), i] as const)`~~. Actually, I don't think this works b/c its in a tuple.

> I also don't fully understand your example - You are pre-filtering the events, then setting the index, then filtering again. The index you receive will be the one after the first filter, so I'm not sure how that helps.

the prefilter is so that the 0th index is the first `ItemCompleted` or `ItemFailed`. doing it this way lets you match up the indices with the indices supplied in the batch call. Versus if you just filter `result.events`, you would still need to offset the indices.

**Comment by ryanleecode** [comment]
If i had the guards then i can do 

```ts
arr.filter((x) => isItemCompleted(x) || x => isItemFailed(x)).map((x, i) => [x, i])
```

which imo would be more ideal
